packages/expo-location/android/src/main/java/expo/modules/location/taskConsumers/LocationTaskConsumer.kt

```diff
@@ -0,0 +1,301 @@
+            mService = null
+          }
+        },
+        Context.BIND_AUTO_CREATE
+      )
+    } else {
+      // Restart the service with new service options.
+      mService!!.startForeground(options.getArguments(FOREGROUND_SERVICE_KEY).toBundle())
+    }
+  }
+
+  private fun stopForegroundService() {
+    val service = mService ?: return
+    service.stop()
+  }
+
+  private fun deferLocations(locations: List<Location>) {
+    val size = mDeferredLocations.size
+    var lastLocation = if (size > 0) mDeferredLocations[size - 1] else mLastReportedLocation
+    locations.forEach { location ->
+      if (lastLocation != null) {
+        mDeferredDistance += abs(location.distanceTo(lastLocation)).toDouble()
+      }
+      lastLocation = location
+    }
+    mDeferredLocations.addAll(locations)
+  }
+
+  private fun maybeReportDeferredLocations() {
+    if (!shouldReportDeferredLocations()) {
+      // Don't report locations yet - continue deferring them.
+      return
+    }
+    val context = context.applicationContext
+    val data: MutableList<PersistableBundle?> = ArrayList()
+    mDeferredLocations.forEach { location ->
+      val timestamp = location.time
+
+      // Some devices may broadcast the same location multiple times (mostly twice) so we're filtering out these locations,
+      // so only one location at the specific timestamp can schedule a job.
+      if (timestamp > sLastTimestamp) {
+        val bundle = locationToBundle(location, PersistableBundle::class.java)
+        data.add(bundle)
+        sLastTimestamp = timestamp
+      }
+    }
+    if (data.size > 0) {
+      // Save last reported location, reset the distance and clear a list of locations.
+      mLastReportedLocation = mDeferredLocations[mDeferredLocations.size - 1]
+      mDeferredDistance = 0.0
+      mDeferredLocations.clear()
+
+      // Schedule new job.
+      taskManagerUtils.scheduleJob(context, mTask, data)
+    }
+  }
+
+  private fun shouldReportDeferredLocations(): Boolean {
+    val task = mTask
+    if (mDeferredLocations.size == 0 || task == null) {
+      return false
+    }
+    if (!mIsHostPaused) {
+      // Don't defer location updates when the activity is in foreground state.
+      return true
+    }
+    val oldestLocation = mLastReportedLocation ?: mDeferredLocations[0]
+    val newestLocation = mDeferredLocations[mDeferredLocations.size - 1]
+    val options: Arguments = MapHelper(task.options)
+    val distance = options.getDouble("deferredUpdatesDistance")
+    val interval = options.getLong("deferredUpdatesInterval")
+    return newestLocation.time - oldestLocation.time >= interval && mDeferredDistance >= distance
+  }
+
+  private fun preparePendingIntent() =
+    taskManagerUtils.createTaskIntent(context, mTask)
+
+  private fun executeTaskWithLocationBundles(
+    locationBundles: ArrayList<Bundle>,
+    callback: TaskExecutionCallback
+  ) {
+    val task = mTask
+    if (locationBundles.size > 0 && task != null) {
```

Comment:
```suggestion
    val task = mTask ?: return 
    if (locationBundles.size > 0) {
```

Or maybe it would be better to create a function `withTask`:
```
inline fun LocationTaskConsumer.withTask(block: TaskInterface -> Unit) {
	mTask?.let {
		block(it)
	}
}
```